file_zbc.c

```diff
@@ -0,0 +1,2240 @@
+			zone->cond = BLK_ZONE_COND_FULL;
+		}
+	}
+
+	return SAM_STAT_GOOD;
+}
+
+/*
+ * Synchronize cache command emulation.
+ */
+static int zbc_flush(struct tcmu_device *dev, struct tcmulib_cmd *cmd)
+{
+	struct zbc_dev *zdev = tcmu_get_dev_private(dev);
+	int ret;
+
+	ret = fsync(zdev->fd);
+	if (ret == 0)
+		ret = zbc_flush_meta(zdev);
+	if (ret) {
+		tcmu_dev_err(dev, "flush failed\n");
+		return tcmu_set_sense_data(cmd->sense_buf,
+					   MEDIUM_ERROR, ASC_WRITE_ERROR, NULL);
+	}
+
+	return SAM_STAT_GOOD;
+}
+
+/*
+ * Handle command emulation.
+ * Return scsi status or TCMU_NOT_HANDLED
+ */
+static int zbc_handle_cmd(struct tcmu_device *dev, struct tcmulib_cmd *cmd)
+{
+	uint8_t *cdb = cmd->cdb;
+	struct iovec *iovec = cmd->iovec;
+	size_t iov_cnt = cmd->iov_cnt;
+	uint8_t *sense = cmd->sense_buf;
+
+	switch (cmd->cdb[0]) {
+
+	case INQUIRY:
+		return zbc_inquiry(dev, cmd);
+
+	case TEST_UNIT_READY:
+		return tcmu_emulate_test_unit_ready(cdb, iovec, iov_cnt, sense);
+
+	case SERVICE_ACTION_IN_16:
+		if (cdb[1] == READ_CAPACITY_16)
+			return zbc_read_capacity(dev, cmd);
+		return TCMU_NOT_HANDLED;
+
+	case MODE_SENSE:
+	case MODE_SENSE_10:
+		return zbc_mode_sense(dev, cmd);
+
+	case MODE_SELECT:
+	case MODE_SELECT_10:
+		return tcmu_emulate_mode_select(dev, cdb,
+						iovec, iov_cnt, sense);
+
+	case ZBC_IN:
+		return zbc_in(dev, cmd);
+
+	case ZBC_OUT:
+		return zbc_out(dev, cmd);
+
+	case READ_6:
+		return TCMU_NOT_HANDLED;
+	case READ_10:
+	case READ_12:
+	case READ_16:
+		return zbc_read(dev, cmd);
+
+	case WRITE_6:
+		return TCMU_NOT_HANDLED;
+	case WRITE_10:
+	case WRITE_12:
+	case WRITE_16:
+		return zbc_write(dev, cmd);
+
+	case SYNCHRONIZE_CACHE:
+	case SYNCHRONIZE_CACHE_16:
+		return zbc_flush(dev, cmd);
+
+	}
+
+	return TCMU_NOT_HANDLED;
+}
+
+static const char zbc_cfg_desc[] =
+	"ZBC emulation device configuration string must be of the form:\n"
+	"\"[opt1[/opt2][...]@]<backstore file path>\n"
+	"Options:\n"
+	"  model-<type>      : Device model. Type must be either HA for\n"
+	"                      host aware or HM for host managed\n"
+	"                      The default is host managed\n"
+	"  lba-<size(B)>     : LBA size in bytes (512 or 4096).\n"
+	"                      The default is 512\n"
+	"  zsize-<size(MiB)> : Zone size in MiB. The default is 256 MiB.\n"
+	"  conv-<num>        : Number of conventional zones at LBA 0 (can be 0)\n"
+	"                      The default is 1%% of the device capacity\n"
+	"  open-<num>        : Optimal (HA) or maximum (HM) number of open zones\n"
+	"                      The default is 128\n"
+	"Ex:\n"
+	"  cfgstring=model-HM/zsize-128/conv-100@/var/local/zbc.raw\n"
+	"  will create a host-managed disk with 128 MiB zones and 100\n"
+	"  conventional zones, stored in the file /var/local/zbc.raw\n";
+
+static struct tcmur_handler zbc_handler = {
+	.cfg_desc = zbc_cfg_desc,
+
+	.name = "ZBC Emulation Handler",
+	.subtype = "zbc",
+
+	.check_config = zbc_check_config,
```

Comment:
The check_config callout was a hack around the kernel/runner not being able to propagate errors at open() time. We are removing this callout, so you should make sure these checks are done at open() time and a -EXYZ error code is returned.